### PR TITLE
419 course search params

### PR DESCRIPTION
**Pull request by goodviber**
### Context
What needs to be done

Wire up the course search flow when searching
By city, town or postcode
Across England
By school, university or other training provider
continue button
Plug the request and query params to make the search so that the age group page can continue the journey
Success

continue button
- validations (see below)
- js vs non js (see below)
js vs non js
- journey still works
- the correct params are passed along
- bookmarks with params displays the options as expected
- 
### Changes proposed in this pull request
Copy over course search from find to replicate params
### Guidance to review
GOTO /find, have a clickathon...

THERE ARE REFERENCES TO SUGGESTED RESULTS IN FIND, NOT IMPLENTED HERE

this PR covers https://trello.com/c/veg6EdVU/581-course-search-initial-flow, https://trello.com/c/FwHmQqrJ/802-course-search-flow-by-school, https://trello.com/c/EXSLehxf/586-course-search-flow-by-city, https://trello.com/c/RTHVaCDs/585-course-search-flow-across-england, https://trello.com/c/L0droAFn/584-course-search-results-page-filters

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased main
- [ ] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/controllers/find/search/locations_controller.rb [hunk 1]
```diff
@@ -1,19 +1,37 @@
...
 
-        if @courses_by_location_or_training_provider_form.valid?
-          redirect_to find_age_groups_path
+        if form_object.valid?
+          parameters_with_geocode_added_and_previous_removed = remove_previous_parameters(form_params.merge(form_object.params))
+          redirect_to(next_step(parameters_with_geocode_added_and_previous_removed))
         else
-          providers
-          render :index
+          flash[:error] = form_object.errors
```
Comment (goodviber) [inline app/controllers/find/search/locations_controller.rb:33]: still need this?

File: app/views/find/result_filters/provider/new.html.erb [hunk 2]
```diff
@@ -10,7 +10,7 @@
         <li data-qa="provider_suggestion">
```
Comment (goodviber) [inline app/views/find/result_filters/provider/new.html.erb:10]: need this?

File: app/views/find/search/locations/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,135 @@
+<%= content_for :page_title, title_with_error_prefix("Find courses by location or by training provider", flash[:error].present?) %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <%= render partial: "find/shared/error_message", locals: {
+      error_anchor_id:
```
Comment (goodviber) [inline app/views/find/search/locations/_form.html.erb:6]: need this?

File: app/views/find/search/subjects/_primary_subjects.html.erb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<div class="govuk-grid-row">
```
Comment (goodviber) [inline app/views/find/search/subjects/_primary_subjects.html.erb:1]: where is this tested?

File: app/helpers/find/subject_helper.rb [hunk 5]
```diff
@@ -20,6 +26,16 @@ def secondary_subject_options(subjects = secondary_subjects)
...
 
+    PrimarySubjectInput = Struct.new(:code, :name)
+
+    def primary_subjects
+      Subject.where(type: "PrimarySubject")
+             .order(:subject_name)
+    end
+
+    # These subjects don’t currently match any courses, and so can be dropped.
+    IGNORED_SUBJECTS = ["Philosophy", "Modern Languages", "Ancient Hebrew", "Ancient Greek"].freeze
```
Comment (goodviber) [inline app/helpers/find/subject_helper.rb:37]: Are theses still to be ignored?

**Review by darokel — COMMENTED** [review]
Review app is looking good particularly the query params matching between new find and current find nice one. I did notice a couple of quirks:

1. Primary search form is missing some copy, see screenshot below:

Screenshot 2022-11-08 at 08 00 34

2. Query params do not persist to the results page once you've selected all your options

Screenshot 2022-11-08 at 08 02 22

3. When searching by school, the query params don't seem to match once you get to the age group form (it seems to add some unnecessary stuff)

PR:

Screenshot 2022-11-08 at 08 04 56

Find:

Screenshot 2022-11-08 at 08 05 12

4. Missing translation for the page title in the primary subjects form

Screenshot 2022-11-08 at 08 07 42

File: app/helpers/find/subject_helper.rb (hunk 5 above)
Comment (defong) [reply app/helpers/find/subject_helper.rb:37]: @avinhurry didn't you got you changed this?

Comment (avinhurry) [reply app/helpers/find/subject_helper.rb:37]: Yes, if you rebase they should be gone 🪄

File: app/views/find/search/locations/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+              <div class="govuk-form-group <%= "govuk-form-group--error" if provider_error? %>" data-module="track-no-provider-results">
+                <!-- if cached providers are present render autocomplete else render sans-js provider search -->
+                <% if select_provider_options(@providers) %>
+                  <% if provider_error? %>
+                     <p class="govuk-error-message" id="provider-error" data-qa="provider-error">
+                     <span class="govuk-visually-hidden">Error: </span><%= flash[:error].last %></p>
+                  <% end %>
+                  <% aria_described_by = "provider-hint" %>
+                  <% aria_described_by += " provider-error" if provider_error? %>
+                  <%#= form.collection_select("query", select_provider_options(@providers), :id, :name, {}, { class: "govuk-input", "aria-describedby" => aria_described_by, data: { qa: "provider-search" } }) %>
```
Comment (defong) [inline app/views/find/search/locations/_form.html.erb:105]: Leftovers

File: app/controllers/find/search/locations_controller.rb [hunk 7]
```diff
@@ -11,6 +11,12 @@ def start; end
       def new; end
 
       def create
+        # if searching by specific provider we have to filter params
+        if provider_option_selected?
+          redirect_to next_step(params)
+          return
```
Comment (darokel) [inline app/controllers/find/search/locations_controller.rb:17]: no need for this return?

Comment (goodviber) [reply app/controllers/find/search/locations_controller.rb:17]: Render and/or redirect were called multiple times in this action. Please note that you may only call render OR redirect, and at most once per action. Also note that neither redirect nor render terminate execution of the action, so if you want to exit an action after redirecting, you need to do something like "redirect_to(...) and return".

File: app/components/find/courses/search_result_component.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
+module Find
```
Comment (goodviber) [inline app/components/find/courses/search_result_component.rb:1]: This is a duplicate to app/components/find_interface/results/search_result_component.rb, plus some other files...

File: app/view_objects/find/results_view.rb [hunk 9]
```diff
@@ -27,12 +27,15 @@ def query_parameters_with_defaults
...
-      @courses ||= RecruitmentCycle.current.courses.includes(
-        :enrichments,
-        subjects: [:financial_incentive],
-        site_statuses: [:site],
-        provider: %i[recruitment_cycle ucas_preferences],
-      ).findable.page(query_parameters[:page] || 1)
+      @courses ||= Find::CourseSearchService.call(
+        filter: query_parameters,
+        sort: query_parameters[:sortby],
+        course_scope: recruitment_cycle.courses,
```
Comment (darokel) [inline app/view_objects/find/results_view.rb:33]: This is incorrect, the scope needs to be findable courses, eg https://github.com/DFE-Digital/publish-teacher-training/blob/main/app/controllers/api/v3/courses_controller.rb#L52. This is the scope that is passed into the current endpoint Find talks to so it needs to match. At the moment the review app returns rolled over courses too

File: app/services/find/course_search_service.rb [hunk 10]
```diff
@@ -0,0 +1,256 @@
+module Find
```
Comment (darokel) [inline app/services/find/course_search_service.rb:1]: I don't think we need to add **_another_** course search service just for Find. Currently, we have the one https://github.com/DFE-Digital/publish-teacher-training/blob/main/app/services/course_search_service.rb which powers the current Find search and the public api search. 

https://github.com/DFE-Digital/publish-teacher-training/blob/main/app/controllers/api/v3/courses_controller.rb#L9 this just delegates to the one above.

**Review by darokel — COMMENTED** [review]
I have been playing around with this locally and found that the result set do not match. Eg I'll make the request on old find and the same on new find for each of the different journeys and the result count is near but still not quite there. I'd expect these to be the same. 

Have provided comments on what I think it is but ultimately the scope being used to search on in the CourseSearch service needs to use findable courses. This PR is just using all courses. 

Another thing, the filtering doesn't seem to be wired up properly. The filter options generate different query parameters which don't seem to match the ones currently in Find. I think the larger issue here is that we have duplicated the search service to have a specific one for find but i don't think we need to as the filter parameters were transformed here https://github.com/DFE-Digital/find-teacher-training/blob/main/app/view_objects/results_view.rb#L440. This method takes the find query params and builds a url which would be accepted by the Publish endpoint. At the moment, the PR doesn't seem to make use of it. 

I think we can resolve most of the search issues by just using the current search service then we could card up making the filters work as sep tickets.

File: app/views/layouts/find_layout.html.erb [hunk 11]
```diff
@@ -40,7 +40,7 @@
         <%= yield %>
       </main>
     </div>
-
+<%= debug params %>
```
Comment (goodviber) [inline app/views/layouts/find_layout.html.erb:43]: to be removed

File: app/components/find/courses/search_result_component.rb (hunk 8 above)
Comment (darokel) [reply app/components/find/courses/search_result_component.rb:1]: Which one do we keep?

File: app/components/find_interface/results/results_component.html.erb [hunk 12]
```diff
@@ -19,7 +19,7 @@
 
     <%= render FindInterface::Results::NoResultsComponent.new(results:) %>
     <%= render FindInterface::Results::SortByComponent.new(results:) %>
-    <%= render FindInterface::Results::SuggestedSearchesComponent.new(results:) %>
+    <%#= render FindInterface::Results::SuggestedSearchesComponent.new(results:) %>
```
Comment (darokel) [inline app/components/find_interface/results/results_component.html.erb:22]: Remnant?

File: app/services/course_search_service.rb [hunk 13]
```diff
@@ -220,7 +242,8 @@ def degree_grades
 
   def subject_codes
     return [] if filter[:subjects].blank?
-    return [] unless filter[:subjects].is_a?(String)
+    # return [] unless filter[:subjects].is_a?(String)
```
Comment (darokel) [inline app/services/course_search_service.rb:245]: remnant?

File: app/view_objects/find/result_filters/filters_view.rb [hunk 14]
```diff
@@ -40,35 +40,42 @@ def provider_query?
...
+
+        params[:study_type].include?("part_time")
       end
 
       def default_study_types_to_true
-        params[:fulltime] != "true" && params[:parttime] != "true"
+        return true if params[:study_type].nil?
+
+        params[:study_type] == "full_time_or_part_time"
+        # params[:fulltime] != "true" && params[:parttime] != "true"
```
Comment (darokel) [inline app/view_objects/find/result_filters/filters_view.rb:74]: remnant?

File: app/views/find/results/_university.html.erb [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+  <ul class="govuk-list govuk-!-margin-top-0 govuk-!-margin-bottom-0">
+    <li>
+      <% if course.further_education? %>
+        <span class="app-description-list__hint govuk-!-margin-top-0">University</span>
+      <% else %>
+        <span class="app-description-list__hint govuk-!-margin-top-0">Placement schools</span>
+        <%= govuk_details(summary_text: @results_view.placement_schools_summary(course)) do %>
+          <p>You can't pick which schools you want to be in, but your university will try to place you in schools you can commute to.</p>
+          <p>Universities usually work with over 100 potential placement schools. Most will be within 10 miles of the university, but sometimes they can cover a wider area, especially outside of cities.</p>
+          <%#= govuk_link_to 'More about placements on this course', course_path(provider_code: course.provider_code, course_code: course.course_code, anchor: 'section-schools') %>
```
Comment (darokel) [inline app/views/find/results/_university.html.erb:13]: remnant?

**Review by darokel — APPROVED** [review]
Left some minor comments but looks good overall (given this is mostly a migration of existing code). Review app and searching also seems to work quite well 👍

File: app/components/find_interface/results/search_result_component.rb [hunk 16]
```diff
@@ -27,12 +27,16 @@ def show_visa_sponsorship_and_degree_required?
...
-          third_class: "An undergraduate degree, or equivalent. This should be an honours degree (Third or above), or equivalent.",
-          not_required: "An undergraduate degree, or equivalent.",
-        }[course.degree_grade.to_sym]
+        if course.degree_grade
+          {
+            two_one: "An undergraduate degree at class 2:1 or above, or equivalent.",
+            two_two: "An undergraduate degree at class 2:2 or above, or equivalent.",
+            third_class: "An undergraduate degree, or equivalent. This should be an honours degree (Third or above), or equivalent.",
+            not_required: "An undergraduate degree, or equivalent.",
+          }[course.degree_grade.to_sym]
```
[the remark this answers is not in the document]
Comment (goodviber) [reply app/components/find_interface/results/search_result_component.rb:36]: We have courses where degree_grade is nil, so I think so?

File: app/view_objects/find/result_filters/filters_view.rb (hunk 14 above)
Comment (goodviber) [reply app/view_objects/find/result_filters/filters_view.rb:74]: yep

File: app/components/find_interface/results/results_component.html.erb (hunk 12 above)
Comment (goodviber) [reply app/components/find_interface/results/results_component.html.erb:23]: TODO

File: app/components/find_interface/results/search_result_component.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (goodviber) [reply app/components/find_interface/results/search_result_component.rb:36]: Sorry I see what you mean `[course.degree_grade&.to_sym || "N/A"]`

File: app/views/find/results/_university.html.erb (hunk 15 above)
Comment (goodviber) [reply app/views/find/results/_university.html.erb:14]: I think it might be TODO

**Pull request merged**